File: cpp/src/ch04/average_score.cc
#include <random>
#include "auto_move_maze_state.h"
#include "hill_climb.h"
#include "random_action.h"
#include "simulated_annealing.h"

void test_ai_score(const StringAIPair &ai, const int game_number)
{
    using std::cout;
    using std::endl;
    std::mt19937 mt_for_constructor(0);
    double score_mean = 0;
    // time it
    auto start = std::chrono::system_clock::now();
    for (int i = 0; i < game_number; i++)
    {
        auto state = State(mt_for_constructor());
        state = ai.second(state);
        auto score = state.get_score(false);
        score_mean += score;
    }
    auto elpased_in_ms = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now() - start).count();
    score_mean /= (double)game_number;
    cout << ai.first << ":\t"
         << score_mean << " elpased: " << elpased_in_ms << "ms" << endl;
}

int main()
{
    int simulate_number = 10000;
    const std::vector<StringAIPair> ais =
        {
            StringAIPair("random", random_action),
            StringAIPair("hill_climb", [&](const State &state)
                         { return hill_climb(state, simulate_number); }),
            StringAIPair("simulated_annealing", [&](const State &state)
                         { return simulated_annealing(
                               state,
                               simulate_number,
                               /* start_temp */ 500,
                               /* end_temp */ 10); })

        };
    int game_number = 100;
    // print what we will do
    using std::cout;
    using std::endl;
    cout << "play " << game_number << " games with " << simulate_number << " simulations" << endl;
    for (const auto &ai : ais)
    {
        cout << ai.first << " ";
    }
    cout << "\n\n";

    // the actual simulation
    for (const auto &ai : ais)
    {
        test_ai_score(ai, game_number);
    }
    return 0;
}

File: python/src/thunder_book/ch04/average_score.py
import random
from typing import Callable

import fire
from tqdm import tqdm

from thunder_book.ch04.auto_move_maze_state import MazeParams
from thunder_book.ch04.auto_move_maze_state import MazeState as State
from thunder_book.ch04.hill_climb import hill_climb
from thunder_book.ch04.random_action import random_action
from thunder_book.ch04.simulated_annealing import simulated_annealing


def test_ai_score(
    name: str,
    action_func: Callable[[State], State],
    game_number: int,
):
    score_mean = 0
    params = MazeParams(width=5, height=5, end_turn=4, num_characters=3)
    for _ in tqdm(range(game_number)):
        state = State(random.randint(0, 100000), params)
        last_state = action_func(state)
        score = last_state.get_score()
        score_mean += score
    score_mean /= game_number
    print(f"score of {name}: {score_mean:.2f}")


type GAMES_TYPE = tuple[str, Callable[[State], State]]


def run(simulate_number=10000, game_number=100):
    games: list[GAMES_TYPE] = [
        ("random_action", random_action),
        ("hill_climb", lambda state: hill_climb(state, simulate_number)),
        (
            "simulated_annealing",
            lambda state: simulated_annealing(
                state,
                simulate_number,
                start_temp=500,
                end_temp=10,
            ),
        ),
    ]

    # print what to play
    print(
        f"average of {game_number} games for {[name for name, _ in games]} in {simulate_number} simulations"
    )

    # play
    for name, action_func in games:
        print(f"play {name}")
        test_ai_score(name, action_func, game_number)


if __name__ == "__main__":
    fire.Fire(run)
